File: ios/Classes/CorePermissionKit/FlutterPermissionKitConfig.swift
//
//  FlutterPermissionKitConfig.swift
//  flutter_permission_kit
//
//  Created by MaoJiu on 2025/5/25.
//

import Foundation

/**
 * PermissionItem
 * 
 * A data structure representing a single permission request with its associated metadata.
 * This struct serves as a bridge between the Flutter layer's permission configuration
 * and the iOS native permission system, encapsulating all necessary information for
 * displaying and processing individual permission requests.
 * 
 * **Purpose**:
 * - Encapsulates permission request data from Flutter layer
 * - Provides optional customization for permission display text
 * - Maps Flutter permission types to iOS PermissionType enum
 * - Enables dynamic permission request configuration
 * 
 * **Data Flow**: Flutter → JSON → Dictionary → PermissionItem → iOS UI
 * 
 * **Example Flutter Configuration**:
 * ```dart
 * PermissionItem(
 *   type: PermissionType.camera,
 *   name: "Camera Access",
 *   description: "Take photos and record videos"
 * )
 * ```
 */
struct PermissionItem {
    /**
     * Optional display name for the permission
     * 
     * Custom title text that will be displayed in the permission card UI.
     * If nil, the permission kit will use its default title (e.g., "Camera", "Photo Library").
     * This allows Flutter developers to customize permission display names based on
     * their app's context and user experience requirements.
     * 
     * **Flutter Mapping**: Corresponds to the `name` field in Flutter's PermissionItem
     * **UI Impact**: Displayed as the main heading in the permission card
     * **Default Behavior**: Uses permission type's default name when nil
     * 
     * **Example**: "Camera Access" instead of default "Camera"
     */
    let name: String?
    
    /**
     * Optional description explaining why the permission is needed
     * 
     * Custom explanatory text that helps users understand why the app needs
     * this specific permission. If nil, the permission kit will use its default
     * description. This is crucial for improving permission grant rates by
     * providing clear, contextual explanations.
     * 
     * **Flutter Mapping**: Corresponds to the `description` field in Flutter's PermissionItem
     * **UI Impact**: Displayed as secondary text below the permission title
     * **Default Behavior**: Uses permission type's default description when nil
     * **Best Practice**: Should be concise but informative about the specific use case
     * 
     * **Example**: "Take photos for your profile and share memories with friends"
     */
    let description: String?
    
    /**
     * The type of permission being requested
     * 
     * Identifies which iOS system permission this item represents (camera, photos, etc.).
     * This is the only required field and determines which permission kit will handle
     * the request and what system APIs will be called.
     * 
     * **Flutter Mapping**: Corresponds to the `type` field in Flutter's PermissionType enum
     * **iOS Integration**: Maps to iOS-specific permission APIs (AVFoundation, Photos, etc.)
     * **Required**: This field must be present and valid for the permission item to be created
     * 
     * **Supported Types**: camera, photos (extensible for future permission types)
     */
    let type: PermissionType
    
    /**
     * Initializes a PermissionItem from a dictionary received from Flutter
     * 
     * This failable initializer parses the dictionary data sent from the Flutter layer
     * via the method channel and creates a strongly-typed PermissionItem instance.
     * The parsing handles optional fields gracefully while ensuring required data is present.
     * 
     * **Data Validation**:
     * - `name` and `description` are optional and can be null/missing
     * - `type` is required and must map to a valid PermissionType
     * - Invalid or missing required data results in nil return
     * 
     * - Parameter dict: Dictionary containing permission data from Flutter
     *   Expected keys: "name" (optional), "description" (optional), "type" (required)
     * 
     * - Returns: A PermissionItem instance if parsing succeeds, nil if required data is missing or invalid
     * 
     * **Expected Dictionary Structure**:
     * ```json
     * {
     *   "name": "Camera Access",           // Optional String
     *   "description": "Take photos...",   // Optional String  
     *   "type": "camera"                   // Required String
     * }
     * ```
     * 
     * **Error Handling**: Returns nil for malformed data, allowing calling code to handle errors gracefully
     */
    init?(from dict: [String: Any]) {
        // Parse optional fields - these can be null from Dart/Flutter
        let name = dict["name"] as? String
        let description = dict["description"] as? String
        
        // Parse required type field and validate it maps to a known PermissionType
        guard let typeString = dict["type"] as? String,
              let type = PermissionType(from: typeString) else {
            // Return nil if type is missing or invalid - this is a critical error
            return nil
        }
        
        self.name = name
        self.description = description
        self.type = type
    }
}

/**
 * FlutterPermissionKitConfig
 * 
 * A comprehensive configuration object that encapsulates all settings and data needed
 * to display and manage the permission request interface. This struct serves as the
 * primary data model for communication between the Flutter layer and iOS native code.
 * 
 * **Architecture Role**:
 * - **Data Transfer Object**: Carries configuration from Flutter to iOS
 * - **UI Configuration**: Contains all text and display settings for the permission interface
 * - **Permission Registry**: Lists all permissions to be requested in the current flow
 * - **Presentation Control**: Determines how the permission UI should be displayed
 * 
 * **Key Features**:
 * - **Type Safety**: Strongly-typed configuration prevents runtime errors
 * - **Flexibility**: Supports custom text for all UI elements
 * - **Extensibility**: Easy to add new configuration options
 * - **Validation**: Built-in parsing validation ensures data integrity
 * 
 * **Usage Flow**:
 * 1. Flutter layer creates configuration object
 * 2. Configuration is serialized and sent via method channel
 * 3. iOS parses dictionary into FlutterPermissionKitConfig
 * 4. Configuration drives permission UI creation and behavior
 */
struct FlutterPermissionKitConfig {
    /**
     * List of permissions to request from the user
     * 
     * An array of PermissionItem objects representing all the permissions that
     * should be displayed and requested in the current permission flow. Each item
     * contains the permission type and optional customization data.
     * 
     * **UI Impact**: Determines how many permission cards are displayed
     * **Processing**: Each item is processed by its corresponding permission kit
     * **Order**: Items are displayed in the order they appear in this array
     * **Validation**: Only items with valid PermissionTypes are included
     * 
     * **Example**: [camera permission, photos permission] for a photo app
     */
    let permissions: [PermissionItem]
    
    /**
     * How the permission request interface should be displayed
     * 
     * Controls the presentation style of the permission request UI, affecting
     * the visual appearance and user interaction patterns. This allows Flutter
     * developers to choose the most appropriate presentation for their app's UX.
     * 
     * **Display Options**:
     * - `.modal`: Full-screen or sheet-style presentation with custom UI
     * - `.alert`: Native iOS alert-style presentation
     * 
     * **UI Impact**: Affects modal presentation style, background, and animations
     * **UX Consideration**: Modal provides more space for explanation, alert is more native
     */
    let displayType: DisplayType
    
    /**
     * The main title displayed at the top of the permission request interface
     * 
     * Primary heading text that introduces the permission request flow to users.
     * This should be clear, concise, and explain the overall purpose of the
     * permission requests that follow.
     * 
     * **UI Location**: Top of the permission interface, prominently displayed
     * **Typography**: Large, bold text that draws attention
     * **Content Guidelines**: Should be action-oriented and user-friendly
     * 
     * **Example**: "Grant Permissions", "App Permissions Required", "Enable Features"
     */
    let displayTitle: String
    
    /**
     * Description text shown in the header section
     * 
     * Explanatory text that appears below the main title and provides context
     * for why the app is requesting permissions. This is crucial for improving
     * user understanding and permission grant rates.
     * 
     * **UI Location**: Below the main title, before the permission cards
     * **Purpose**: Explains the overall benefit or necessity of granting permissions
     * **Length**: Should be concise but informative (1-2 sentences ideal)
     * **Tone**: Friendly and transparent about permission usage
     * 
     * **Example**: "To provide the best experience, this app needs access to the following features."
     */
    let displayHeaderDescription: String
    
    /**
     * Description text shown at the bottom of the permission list
     * 
     * Additional explanatory or reassuring text that appears after all permission
     * cards. Often used for privacy assurances, additional context, or next steps
     * information.
     * 
     * **UI Location**: Below all permission cards, at the bottom of the interface
     * **Purpose**: Provides final context, reassurance, or instructions
     * **Common Uses**: Privacy statements, "You can change these later" messages
     * **Optional Content**: Can be empty if no additional context is needed
     * 
     * **Example**: "You can modify these permissions anytime in Settings."
     */
    let displayBottomDescription: String
    
    /**
     * Initializes a FlutterPermissionKitConfig from a dictionary received from Flutter
     * 
     * This failable initializer parses the complete configuration dictionary sent
     * from Flutter and creates a strongly-typed configuration object. It handles
     * nested data structures (permissions array) and validates all required fields.
     * 
     * **Parsing Process**:
     * 1. Extracts and validates permissions array
     * 2. Parses each permission item, filtering out invalid entries
     * 3. Validates display type string and converts to enum
     * 4. Extracts all required text fields
     * 5. Returns nil if any required field is missing or invalid
     * 
     * - Parameter dict: Dictionary containing complete configuration from Flutter
     *   Must contain all required fields with correct types
     * 
     * - Returns: A FlutterPermissionKitConfig instance if parsing succeeds, nil otherwise
     * 
     * **Expected Dictionary Structure**:
     * ```json
     * {
     *   "permissions": [
     *     {"type": "camera", "name": "Camera", "description": "..."},
     *     {"type": "photos", "name": "Photos", "description": "..."}
     *   ],
     *   "displayType": "modal",
     *   "displayTitle": "Grant Permissions",
     *   "displayHeaderDescription": "This app needs...",
     *   "displayBottomDescription": "You can change..."
     * }
     * ```
     * 
     * **Error Handling**: Returns nil for malformed data, allowing graceful error handling
     * **Data Integrity**: Validates all fields to prevent runtime errors in UI components
     */
    init?(from dict: [String: Any]) {
        // Parse and validate the permissions array
        guard let permissionsArray = dict["permissions"] as? [[String: Any]] else {
            return nil
        }
        
        // Convert permission dictionaries to PermissionItem objects, filtering out invalid entries
        let permissions = permissionsArray.compactMap { PermissionItem(from: $0) }
        
        // Parse and validate all other required configuration fields
        guard let displayTypeString = dict["displayType"] as? String,
              let displayType = DisplayType(from: displayTypeString),
              let displayTitle = dict["displayTitle"] as? String,
              let displayHeaderDescription = dict["displayHeaderDescription"] as? String,
              let displayBottomDescription = dict["displayBottomDescription"] as? String
        else {
            return nil
        }
        
        // Initialize with validated data
        self.permissions = permissions
        self.displayType = displayType
        self.displayTitle = displayTitle
        self.displayHeaderDescription = displayHeaderDescription
        self.displayBottomDescription = displayBottomDescription
    }
}

File: lib/core/flutter_permission_kit_config.dart
import 'package:flutter_permission_kit/core/permission.dart';
import 'package:flutter_permission_kit/enums/display_type.dart';

/// Configuration class for Flutter Permission Kit
///
/// This class contains all the configuration options for customizing the
/// appearance and behavior of permission request dialogs. It allows you to
/// specify which permissions to request, how they should be displayed,
/// and various UI customization options.
///
/// Example usage:
/// ```dart
/// // Using custom permission names and descriptions
/// final config = FlutterPermissionKitConfig(
///   permissions: [
///     Permission(
///       name: 'Camera',
///       description: 'Take photos and videos',
///       type: PermissionType.camera,
///     ),
///   ],
///   displayType: DisplayType.modal,
///   displayTitle: 'App Permissions',
/// );
///
/// // Using convenient static methods (name and description will be null)
/// final quickConfig = FlutterPermissionKitConfig(
///   permissions: [
///     Permission.camera,
///     Permission.photos,
///     Permission.microphone,
///   ],
///   displayType: DisplayType.alert,
///   displayTitle: 'Required Permissions',
/// );
/// ```
class FlutterPermissionKitConfig {
  /// List of permissions to request from the user
  ///
  /// This is a required list containing all the permissions that should be
  /// requested in this permission flow. Each Permission object contains
  /// the permission type and optional display information.
  final List<Permission> permissions;

  /// How the permission request should be displayed to the user
  ///
  /// Determines whether to show an alert dialog or a full-screen modal.
  /// Defaults to [DisplayType.alert] for a more native feel.
  final DisplayType displayType;

  /// The main title displayed at the top of the permission request
  ///
  /// This is the primary heading that users will see when the permission
  /// request appears. Should be concise and descriptive.
  /// Defaults to 'Permission Request'.
  final String displayTitle;

  /// Description text shown in the header section
  ///
  /// This appears below the title and provides context about why
  /// permissions are being requested. Can be used to explain the
  /// overall purpose or benefit of granting the permissions.
  /// Defaults to an empty string.
  final String displayHeaderDescription;

  /// Description text shown at the bottom of the permission list
  ///
  /// This appears after all permission items and can be used for
  /// additional information, privacy notes, or reassurances about
  /// data usage. Defaults to an empty string.
  final String displayBottomDescription;

  /// Creates a new FlutterPermissionKitConfig instance
  ///
  /// [permissions] is required and must contain at least one permission.
  /// All other parameters are optional and have sensible defaults.
  ///
  /// The default configuration provides a good starting point for most apps:
  /// - Alert-style display for native feel
  /// - iOS-style blue primary color
  /// - Auto-dismiss and auto-check enabled for better UX
  FlutterPermissionKitConfig({
    required this.permissions,
    this.displayType = DisplayType.alert,
    this.displayTitle = 'Permission Request',
    this.displayHeaderDescription =
        'To provide key features and a seamless experience, we require access to certain permissions.',
    this.displayBottomDescription =
        'Permissions can be managed later in Settings if you change your mind.',
  });

  /// Converts this FlutterPermissionKitConfig instance to a JSON-serializable Map
  ///
  // ignore: unintended_html_in_doc_comment
  /// This method serializes the entire configuration object into a Map<String, dynamic>
  /// format suitable for JSON conversion. This is essential for passing configuration
  /// data to native platform code through method channels, storing settings in files,
  /// or transmitting configuration over network connections.
  ///
  /// The serialization process handles complex data types appropriately:
  /// - Permission objects are converted using their individual toJson() methods
  /// - Enum values are converted to their string representations
  /// - Color objects are converted to their string representation
  /// - Primitive types (String, bool) are preserved as-is
  ///
  /// Returns a Map containing all configuration properties:
  /// - 'permissions': Array of serialized Permission objects
  /// - 'displayType': String representation of the DisplayType enum (e.g., "alert", "modal")
  /// - 'displayTitle': The main title text for the permission dialog
  /// - 'displayHeaderDescription': Header description text explaining the permission request
  /// - 'displayBottomDescription': Footer description text with additional information
  /// - 'primaryColor': String representation of the Color object (e.g., "Color(0xff007aff)")
  /// - 'autoDismiss': Boolean indicating whether to auto-dismiss the dialog
  /// - 'autoCheck': Boolean indicating whether to auto-check permission status
  ///
  /// Example output:
  /// ```json
  /// {
  ///   "permissions": [
  ///     {
  ///       "name": "Camera Access",
  ///       "description": "Take photos and videos",
  ///       "type": "camera"
  ///     }
  ///   ],
  ///   "displayType": "alert",
  ///   "displayTitle": "App Permissions",
  ///   "displayHeaderDescription": "We need these permissions...",
  ///   "displayBottomDescription": "You can change these later...",
  /// }
  /// ```
  ///
  /// This serialized format ensures that all configuration data can be reliably
  /// transmitted to platform-specific code and reconstructed as needed.
  Map<String, dynamic> toJson() => {
    'permissions': permissions.map((e) => e.toJson()).toList(),
    'displayType': displayType.name,
    'displayTitle': displayTitle,
    'displayHeaderDescription': displayHeaderDescription,
    'displayBottomDescription': displayBottomDescription,
  };
}
